cmd/transform/transform.go

```diff
@@ -0,0 +1,162 @@
+	"github.com/konveyor/crane-lib/transform"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+)
+
+type Options struct {
+	logger       logrus.FieldLogger
+	ExportDir    string
+	PluginDir    string
+	TransformDir string
+}
+
+func (o *Options) Complete(c *cobra.Command, args []string) error {
+	// TODO: @sseago
+	return nil
+}
+
+func (o *Options) Validate() error {
+	// TODO: @sseago
+	return nil
+}
+
+func (o *Options) Run() error {
+	return o.run()
+}
+
+func NewTransformCommand() *cobra.Command {
+	o := &Options{
+		logger: logrus.New(),
+	}
+	cmd := &cobra.Command{
+		Use:   "transform",
+		Short: "Create the transformations for the exported resources and plugins and save the results in a transform directory",
+		RunE: func(c *cobra.Command, args []string) error {
+			if err := o.Complete(c, args); err != nil {
+				return err
+			}
+			if err := o.Validate(); err != nil {
+				return err
+			}
+			if err := o.Run(); err != nil {
+				return err
+			}
+
+			return nil
+		},
+	}
+
+	addFlagsForOptions(o, cmd)
+
+	return cmd
+}
+
+func addFlagsForOptions(o *Options, cmd *cobra.Command) {
+	cmd.Flags().StringVarP(&o.ExportDir, "export-dir", "e", "export", "The path where the kubernetes resources are saved")
+	cmd.Flags().StringVarP(&o.PluginDir, "plugin-dir", "p", "plugins", "The path where binary plugins are located")
+	cmd.Flags().StringVarP(&o.TransformDir, "transform-dir", "t", "transform", "The path where files that contain the transformations are saved")
+}
+
+func (o *Options) run() error {
+	runner := transform.Runner{}
+
+	// Load all the resources from the export dir
+	exportDir, err := filepath.Abs(o.ExportDir)
+	if err != nil {
+		// Handle errors better for users.
+		return err
+	}
+
+	pluginDir, err := filepath.Abs(o.PluginDir)
+	if err != nil {
+		return err
+	}
+
+	transformDir, err := filepath.Abs(o.TransformDir)
+	if err != nil {
+		return err
+	}
+
+	plugins, err := plugin.GetBinaryPlugins(pluginDir)
+	if err != nil {
+		return err
+	}
+
+	files, err := file.ReadFiles(context.TODO(), exportDir)
+	if err != nil {
+		return err
+	}
+
+	opts := file.PathOpts{
+		TransformDir: transformDir,
+		ExportDir:  exportDir,
+	}
+
+	runner = transform.Runner{}
+
+	for _, f := range files {
+		patches, haveWhiteOut, err := runner.Run(f.Unstructured, plugins)
+		if err != nil {
+			return err
+		}
+
+		if haveWhiteOut {
+			whPath := opts.GetWhiteOutFilePath(f.Path)
+			_, statErr := os.Stat(whPath)
+			if os.IsNotExist(statErr) {
+				o.logger.Infof("resource file: %v creating whiteout file: %v", f.Info.Name(), whPath)
+				err = os.MkdirAll(filepath.Dir(whPath), 0700)
+				if err != nil {
+					return err
+				}			
+				whFile, err := os.Create(whPath)
+				if err != nil {
+					return err
+				}
+				whFile.Close()
+			}
+			continue
+		} else {
+			// if whiteout file exists from prior run, remove it
+			whPath := opts.GetWhiteOutFilePath(f.Path)
+			_, statErr := os.Stat(whPath)
+			if !os.IsNotExist(statErr) {
```

Comment:
What it would look like to instead overwrite the entire transform dir every time. 

I think that doing this complicated logic to try and keep old changes and such is going to be pretty error-prone.

Reply:
Are you suggesting that we recursively delete the transform dir and its contents prior to writing any files?

Reply:
But to be clear, the logic here does not attempt to keep old changes. The intent is to make the whiteout and patch files look like they should for the current run. For whiteout, the two questions are 1) does the file exist, and 2) should the file exist? If these two questions have different answers, then either create or delete the file as appropriate. Arguably, we could skip the "does it exist? question when we want it to exist and just always issue the create call. That would avoid having to check for prior existence in the whiteout case. For haveWhiteOut==false, though, I think we still need to check for existence and only delete if it exists from before. In the case of the transform patches file, it's simpler. We always write a file (even if it's zero length), so we always just open the file with `os.Create`, meaning we always get a new file, overwriting any prior files. The only difference in outcome if we nuke the top-level transform dir ahead of time is that if there a re files there that do not correspond to resources in the export dir, those will also be deleted. I'm assuming that apply will ignore these files if they aren't deleted anyway, so this probably wouldn't have any effect on the outcome.

Reply:
They may not have an effect on the outcome, but I am wondering if that is a bad user expectation, I think for now if we do the TODO's that you outlined that this is acceptable and once we talk the next step with transform's then we will flush these out